R/cnd-abort.R

```diff
@@ -248,6 +290,345 @@ signal_abort <- function(cnd, file = NULL) {
+#' to climb the call stack one level to find the relevant call. In the
+#' following example, the complexity is not so important that sparing
+#' the argument passing makes a big difference. However this
+#' illustrates the pattern:
+#'
+#' ```
+#' arg_check <- function(arg,
+#'                       error_arg = as_string(substitute(arg)),
+#'                       error_call = caller_env()) {
+#'   # Set the local error call
+#'   local_error_call(error_call)
+#'
+#'   my_classed_stop(
+#'     cli::format_error("{.arg {error_arg}} is failing.")
+#'   )
+#' }
+#'
+#' my_classed_stop <- function(message) {
+#'   # Forward the local error call to the caller's
+#'   local_error_call(caller_env())
+#'
+#'   abort(message, class = "my_class")
+#' }
+#'
+#' foo <- function(x) arg_check(x)
+#' foo()
+#' #> Error in `foo()`: `x` is failing.
+#' ```
+#'
+#' @section Error call flags in performance-critical functions:
+#'
+#' The `call` argument can also be the string `"caller"`. This is
+#' equivalent to `caller_env()` or `parent.frame()` but has a lower
+#' overhead because call stack introspection is only performed when an
+#' error is triggered. Note that eagerly calling `caller_env()` is
+#' fast enough in almost all cases.
+#'
+#' If your function needs to be really fast, assign the error call
+#' flag directly instead of calling `local_error_flag()`:
+#'
+#' ```
+#' .__error_call__. <- "caller"
+#' ```
+#' 
+#' @examples
+#' # Set a context for error messages
+#' function() {
+#'   local_error_call(quote(foo()))
+#'   local_error_call(sys.call())
+#' }
+#'
+#' # Disable the context
+#' function() {
+#'   local_error_call(NULL)
+#' }
+#'
+#' # Use the caller's context
+#' function() {
+#'   local_error_call(caller_env())
+#' }
+#' @export
+local_error_call <- function(call, frame = caller_env()) {
+  if (!is_string(call, "caller") && !typeof(call) %in% c("environment", "NULL", "language")) {
+    abort(sprintf("`call` can't be %s.", friendly_type_of(call)))
+  }
+
+  old <- frame$.__error_call__.
+  frame$.__error_call__. <- call
+
+  old
+}
+
+#' Documentation anchor for error arguments
+#'
+#' @description
+#'
+#' Use `@inheritParams rlang::args_error_context` in your package to
+#' document `arg` and `call` arguments (or equivalently their prefixed
+#' versions `error_arg` and `error_call`).
+#'
+#' - `arg` parameters should be formatted as argument (e.g. using
+#'   cli's `.arg` specifier) and included in error messages. See also
+#'   [caller_arg()].
+#'
+#' - `call` parameters should be included in error conditions in a
+#'   field named `call`. An easy way to do this is by passing a `call`
+#'   argument to [abort()]. See also [local_error_call()].
+#'
+#' @param arg,error_arg An argument name as a string. This argument
+#'   will be mentioned in error messages as the input that is at the
+#'   origin of a problem.
+#' @param call,error_call An expression (as returned by e.g.
```

Comment:
Should it usually be an environment? In most checker functions it'll be `caller_env()`, right? And then supplying a call is just a special case?